Scope users in sites

entantoencuanto opened the pull request:
Related with #1312

### What does this PR do?
* Renames the source_site attribute of users with site.
* Changes the association between `User` and `Site` adding a `has_many :users, dependent: :nullify`.
* Removes the unique database index for `email` on users and creates a new one unique for `email` and `site_id`.
* Defines a validation of uniqueness of `email` scoped to site in User model.
* In `sessions_controller` a user will only be able to sign in in their site. This is implemented by including the site in the `User::SessionsForm`.
* Refactor tests taking into account the site and includes some test:
  * Verifies a user can't use email/password to sign in in wrong site.
  * Verifies the list of users in admin/users only displays users of the site.
  * Test that `User::SessionsForm` must include the site to be valid.


### How should this be manually tested?
Register with the same email in different sites, try to sign in with the credentials of one site in other. As admin user take a look to the list of users of a site.

### Does this PR changes any configuration file?

No

entantoencuanto committed:
#1312: Rename source_site reference with site in users table

entantoencuanto committed:
#1312: Set unique index for email and site_id

entantoencuanto committed:
#1312: Rename User scope by_source_site with by_site

entantoencuanto committed:
Update db/schema.rb

entantoencuanto committed:
#1312: Rename calls to source_site with site in User instances

entantoencuanto committed:
#1312: Change association between Site and User

entantoencuanto committed:
#1312: Set uniqueness of users email scoped to site

entantoencuanto committed:
#1312: Consider the site in default user authentication

entantoencuanto committed:
#1312: Raise an exception if with_current_site helper is called nested inside other

entantoencuanto committed:
#1312: Consider the site un with_signed_in_user helpers in tests

entantoencuanto committed:
#1312: Change and extend sessions_controller test

entantoencuanto committed:
#1312: Avoid the nesting of with_current_site and with_signed_in_user in tests

entantoencuanto committed:
#1312: Add presence validation of site in User::SessionForm

entantoencuanto committed:
#1312: Update User::SessionForm test

entantoencuanto committed:
#1312: Fix some tests using appropriate unverified user on corresponding site

entantoencuanto committed:
#1312: Replace some tests users with fitting site

entantoencuanto committed:
#1312: Fix test defining current_site

entantoencuanto committed:
#1312: Update test after fixtures addition of census verification

entantoencuanto committed:
#1312: Update docs/user-namespace.md

entantoencuanto committed:
#1312: Test that admin/users only shows users of site

entantoencuanto committed:
#1312: Add missing translations

ferblape reviewed (changes requested):
I have created a new user in burjassot.gobify.net and automatically I got logged in dival.gobify.net and the rest of gobify.net domains. You should limit the cookie domain to the site domain for users

entantoencuanto committed:
#1312: Consider site in user email confirmation process

The user will be searched by token with the site given by
the request. In this way there is only one valid confirmation url.

entantoencuanto committed:
#1312: Test confirmations_controller. Check valid and invalid site

entantoencuanto committed:
#1312: Reorder arguments on ConfirmationForm initializer to avoid errors

entantoencuanto committed:
#1312: Include site in ConfirmationForm tests

entantoencuanto committed:
#1312: Correct translation language

entantoencuanto committed:
#1312: find_current_user by user_id only if user belongs to current_site

entantoencuanto committed:
#1312: Disable all domains option on session_store

entantoencuanto committed:
#1312: Add helper method to tests to include the host of the site in the session

entantoencuanto committed:
#1312: Add tests to check that a user can only manage session in his host

entantoencuanto committed:
#1312: Include users site host in requests of with_signed_in_user

entantoencuanto committed:
#1312: Fix with_current_site method to not set app_host

entantoencuanto committed:
#1312: Remove trailing char from host in with_site_host

entantoencuanto committed:
#1342: Set current_site from request in admin if admin has permission

entantoencuanto committed:
#1342: Avoid infinite redirection loops signing out from admin

entantoencuanto committed:
#1342: Avoid redirection when admin changes site of request

entantoencuanto committed:
#1312: Add some tests to check admin sessions depending on host

entantoencuanto committed:
Update db/schema.rb

entantoencuanto committed:
#1312: Fix test nesting with_signed_in_user in with_current_site

entantoencuanto commented:
Last commits implements the requested change. The users session won't remain between different sites, also admins sessions.

Also, some commits has been added to fix #1342 and some redirection errors signing out from admin session.

entantoencuanto committed:
#1312: Add test to check that a user can use same email to register in different sites

entantoencuanto committed:
#1312: Recover session_store domain: :all configuration

entantoencuanto committed:
#1312: Change tests reflecting the session is shared between domains

entantoencuanto committed:
#1312: Add test to check uniqueness of email scoped to site in User

ferblape reviewed (approved):
🚀
